### ISSUE TITLE
Failed to load an image. Image was too large

### ISSUE THREAD

**Issue by metal450** [FFDA/SourCherry#11]
A bunch of my notes are inaccessible, as accessing them shows "Failed to load an image.  Image was too large."  Any way to get rid of that limitation?  All of the images are as links/attachments, not as direct image embeds.

**Comment by FFDA** [comment]
This error message might be displayed by image, attached file, anchor or latex box. I don't think latter two can do it, it most likely is caused by attached files/images. I just made it say it's an image because I encountered it with images first.

> Any way to get rid of that limitation

Not that I know off. There are no mentions in official documentation (or I couldn't find any), but I found mentions on stackoverflow that Google's sqlite3 implementation is limited compared to standard/normal one.

I tried to figure out limit of the image that can be attached to the node, but couldn't get exact size. While searching on the internet you can find that sqlite cursor limit is 1MB, but it's definitely lower while trying to display image (the same goes for files). In my experience anything over 500kB will not load. But it can go as low as 400kB.

Try to open the same node in XML database. If I recall correctly I did not have these type of issues with that type of database.

**Comment by FFDA** [comment]
Which android version your device is running? If your phone runs on Android 9 or later [try this apk](https://github.com/FFDA/ffda.github.storage/raw/main/SourCherry-releases/SourCherry_1.0.0-bigger_cursorWindow.apk).

**Comment by metal450** [comment]
I'm on Android 9.

That build is MUCH better! Now the message is "Embedded file/image was too big (>15mb), but almost all of the ones that didn't work before now do - only a tiny minority of attachments are that large, whereas before, probably half the attachments didn't work :)

One minor suggestion: I noticed that in the case where that message appears & the attachment is inaccessible, it just shows nothing in the note content to indicate the missing attachment (I had to launch CherryTree on PC to see what was there). Perhaps it could still show the attachment name in-place in the note content, so we can see what's missing, but have something like a "broken link" icon or something. Just to indicate where/what it is that's inaccessible?

In any event, this build definitely improved it a lot.

**Comment by FFDA** [comment]
What kind of files are you attaching that they are so big? It's a note taking app and not some kind of file storage and you are loosing a lot of space. As an example, 497.5KiB image encoded to base64 in CherryTree SQL database can take as much as 2.8MB. That's what happened for me when I noticed that issue.

> Perhaps it could still show the attachment name in-place in the note content
 
It is possible to do so. For attached files it is still possible to display filenames. For images - a small placeholder.

I think I will look into adding a setting where user can adjust cursor window size. Because any issues caused by attached files larger than 10MB won't be addressed by me.

**Comment by metal450** [comment]
> It's a note taking app and not some kind of file storage

Yes, I'm aware that it's a note taking app and not some kind of file storage.

> What kind of files are you attaching that they are so big? 

In this case it was a zip that had a few of related images. I can unzip & attach the images separately (I'm also aware that zip doesn't further compress compressed images, it was just for ease/quickness vs attaching the images 1 at a time, as it worked ok in the PC version).

> I think I will look into adding a setting where user can adjust cursor window size.

Just curious, is there any downside to just having a huge cursor size? i.e. does that have any negative impact? If no downside, seems like might as well just forego a setting & have it be as big as possible?

Anyway, without addressing whether or not I'm "correctly" using CherryTree, and regardless of setting/limit, would still probably be nice to see a placeholder of what it's not loading - while 10-15mb is more than enough for 99.99% of cases, as long as the behavior is more "strict" than the PC version, someone will likely encounter it from time to time.  Showing the problematic file can tell the user exactly what they should go to their PC & remove or fix.

**Comment by FFDA** [comment]
The only thing that I can think off is app/phone running out of memory. Android might decide that memory usage is excessive and simply kill the app.
And I do not know if it would do it even when one opens a 500MB file, but there is a limit. I haven't read through all the Android documentation, but I know that it is written multiple times that app might be killed at any time, especially in the background, to save/free memory for other task, however app being in the foreground does not guarantee that it will not be killed.
So, there is a chance that to open/save bigger files I would have to do it in some different way that requires a lot of code and research. Starting Monday I might not have a lot of time left to work on the app and the next thing I want to focus is DB editing. Something simple, like renaming/deleting nodes. Maybe even plain text node creation and editing.

**Comment by metal450** [comment]
Gotcha. Anyway yeah, I think the current limit is probably fine. I only found one or two nodes now that work on PC but not Android, and they're the extreme minority (whereas previously, it was like 20% of them, as just regular jpegs were over the limit).  So aside from showing the name of blocked attachments, I think the current limit is good enough :)

**Comment by FFDA** [comment]
With 1.1.0 release I added a setting to change SQL cursor window. Default is 15MB, but I think it can be increased more without any issues. I think I tried 30+MB file (window size set to 50MB) and app didn't crash and there were no errors. However, XML database had issues saving that file, it crashed. I don't know exactly how to solve it, because it's OOM error I can't even catch it. I think the biggest issue is that XML parser I use loads all the document in the memory and that leaves too little memory for anything else. I could be wrong, but the only way to make sure is use a different XML parser and that's a lot of work. I would have to rewrite a lot of if not all XMLReader class.
In conclusion I won't do anything else for XML databases for now and wait for someone to make another issue. I will close this Issue in couple of days. Or you can do it too.

**Comment by metal450** [comment]
Yup, I saw it - and no issues for me personally, I only use SQL (as you mention, xml is far too inefficient/memory-hungry, not really a format well suited to databases like this)  Anyway, all good - thanks!
